neps/nep-0452.md

```diff
@@ -0,0 +1,221 @@
+- the linkdrop contract will delete the access key associated with `pubKey1` so that it cannot be used again.
+- the linkdrop contract will create the new account and transfer the funds to it alongside any other assets.
+- the user will be able to sign transactions on behalf of the new account using `privKey2`.
+
+### Claiming a linkdrop with a NEAR Account
+
+A user with an *existing* wallet that is claiming the assets associated with the public key:
+
+- signs a transaction on behalf of the linkdrop contract (contract is set to signer_id) using `privKey1`. The method will be `claim`.
+- the args of this function call will simply contain the user's existing account ID.
+- the linkdrop contract will delete the access key associated with `pubKey1` so that it cannot be used again.
+- the linkdrop contract will transfer the funds to that account alongside any other assets.
+
+
+## Reference Implementation
+
+Below are some references for linkdrop contracts.
+
+- [Link Drop Contract](https://github.com/near/near-linkdrop)
+
+- [Keypom Contract](https://github.com/keypom/keypom)
+
+
+## Security Implications
+
+1. Linkdrop Creation
+Linkdrop creation involves creating keypairs associated with mapping to assets such as $NEAR, FTs, NFTs and access keys. There are security implications with the access keys associated with the keypairs. These access keys should be function call access keys and be limited to specific functionality. For example, keypairs can be generated with function call access keys which have permission to only call `claim` and `create_account_and_claim`. Another important security implication of linkdrop creation is to ensure that only one key is mapped to a set of assets at any given time. Linkdrop creation can happen in real-time or prior to being received. The mass creation of linkdrops involves creating multiple keypairs associated with mapping to assets and access keys. The security implications mentioned above are increased with the mass creation of linkdrops.
+
+2. Linkdrop Key Management  
+Key management is a critical safety component of linkdrops. The linkdrop creator should implement a key management strategy for linkdrop keys upon the receiver claiming the linkdrops assets. For example, one strategy may be to delete the linkdrop keypair in the same block the receiver claims the linkdrop assets, so a single linkdrop cannot be used multiple times. In another example, a linkdrop creator may allow for a linkdrop to be used X number of times, where they could track the number of calls to `claim` or `create_account_and_claim` and delete the keypair after X uses of the linkdrop. It is very important to implement your key management strategy in `claim` and `create_account_and_claim` as opposed to a callback function to ensure the linkdrop keys are not improperly used. Implementing key management in a callback has a security implication of such as 'double-spending' a linkdrop.
```

Comment:
> one strategy may be to delete the linkdrop keypair in the same block the receiver claims the linkdrop assets

This is impossible to do in NEAR as the key deletion receipt produced by the `claim` function call will not be included, let alone, executed in the block where `claim` function call receipt is processed. The assets MUST be locked - either removed from KeyInfo and re-added in the callback on error OR marked "in transfer" in some other way.